### PR TITLE
Nonlinear state spaces

### PR DESCRIPTION
**Pull request by cpfiffer**
This PR adds the capacity for DifferenceEquations.jl to handle arbitrary evolution, volatility, and observation functions. There's been some minor cleanup and the noise types are a little more rational, though could still use some work.

Tests are en route, I just wanted to get this on people's radar for now. Seems to be reasonably fast, haven't found any type instability yet. 

Here's an example that works with DifferentiableStateSpaceModels.jl (with some modification):

```julia
using DifferenceEquations
using DifferentiableStateSpaceModels
using DifferentiableStateSpaceModels.Examples

# Grab the model
m = @include_example_module(Examples.rbc_observables)
p_f = (ρ=0.2, δ=0.02, σ=0.01, Ω_1=0.1)
p_d = (α=0.5, β=0.95)

# Generate cache, create perutrbation solution
c = SolverCache(m, Val(1), p_d)
sol = generate_perturbation(m, p_d, p_f; cache = c)

# Timespan to simulate across
T = 500

# Set initial state
u0 = zeros(m.n_x)

# Construct problem with no observables
problem = StateSpaceProblem(
    DifferentiableStateSpaceModels.dssm_evolution,
    DifferentiableStateSpaceModels.dssm_volatility,
    DifferentiableStateSpaceModels.dssm_observation,
    u0,
    (1,T),
    sol
)

# Solve the model, this generates
# simulated data.
simul = DifferenceEquations.solve(problem)

# Extract the observables, latent noise, and latent states.
z, n, u = simul.z, simul.n, simul.u

# Now solve using the previous data as observables.
# Solving this problem also includes a likelihood.
problem_data = StateSpaceProblem(
    DifferentiableStateSpaceModels.dssm_evolution,
    DifferentiableStateSpaceModels.dssm_volatility,
    DifferentiableStateSpaceModels.dssm_observation,
    u0,
    (1,T),
    sol,
    observables = z
)

# Generate likelihood.
s2 = DifferenceEquations.solve(problem_data)
```

### CODE REVIEW

File: src/linear.jl [hunk 1]
```diff
@@ -69,18 +69,12 @@ function LinearStateSpaceProblem(
...
-function SciMLBase.solve(
+CommonSolve.init(::LinearStateSpaceProblem, args...; kwargs...) = LinearGaussian()
+solve!(::LinearGaussian) = StateSpaceSolution(missing,missing,missing,missing) # TODO: Need more sensible defaults
+
+function CommonSolve.solve!(
     prob::LinearStateSpaceProblem{isinplace, Atype, Btype, Ctype, wtype, Rtype, utype, ttype, otype}, 
-    ::LinearGaussian, 
     args...; 
     kwargs...
 ) where {isinplace, Atype, Btype, Ctype, wtype, Rtype<:AbstractMatrix, utype, ttype, otype<:Nothing}
```
Comment (cpfiffer) [inline src/linear.jl:80]: Pretty sure this is not the right implementation of the CommonSolve interface. Should it instead be `init` passes the problem to the solver and then `solve!` is called on the solver, i.e. I'd use something like `alg.prob.stuff` to get fields from the problem type?

File: cannon.jl [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+prob1 = LinearStateSpaceProblem(A, B, C, u0, tspan, R=R)
+sol1 = solve(prob1, LinearGaussian())
+
+prob2 = LinearStateSpaceProblem(A, B, C, u0, tspan, R=R, observables=sol1.z)
+sol2 = solve(prob2, LinearGaussian())
+
+prob3 = LinearStateSpaceProblem(A, B, C, u0, tspan, R=R, observables=sol2.z, noise=DefinedNoise(sol2.n))
+sol3 = solve(prob3, LinearGaussian())
+
+@assert sol2.n == sol3.n
```
Comment (ChrisRackauckas) [inline cannon.jl:31]: Is this supposed to be a test? Add it to the tests?

File: integ/go.jl [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    (1,T),
+    sol
+)
+
+
+simul = DifferenceEquations.solve(problem)
+
+z, n, u = simul.z, simul.n, simul.u
+
+problem_data = StateSpaceProblem(
```
Comment (ChrisRackauckas) [inline integ/go.jl:32]: same here.

File: src/linear.jl (hunk 1 above)
Comment (ChrisRackauckas) [reply src/linear.jl:80]: `init` passes to a `cache` or `integrator` object on which `solve!` is called.

Comment (ChrisRackauckas) [reply src/linear.jl:80]: This allows for reusing the same cache across different solves if someone `reinit!` to a new `u0` or parameters.

File: src/noise.jl [hunk 4]
```diff
@@ -1,15 +1,34 @@
...
 struct DefinedNoise{T} <: AbstractNoise
     values::T
 end
 
 noise(s::DefinedNoise{<:Vector}, t) = s.values[t]
+
+# Likelihood definitions
+StatsBase.loglikelihood(x, noise::Gaussian{<:Any, <:Nothing}, t) = loglikelihood(MvNormal(diagm(ones(noise.size))), x)
+StatsBase.loglikelihood(x, noise::Gaussian, t) = loglikelihood(MvNormal(noise.B), x)
+StatsBase.loglikelihood(x, noise::AbstractMatrix, t) = loglikelihood(MvNormal(noise), x)
```
Comment (ChrisRackauckas) [inline src/noise.jl:34]: Is there a reason this cannot or should not use the DiffEqNoiseProcess interface? It looks like the likelihoods would just need to be added on there.

File: src/nonlinear.jl [hunk 5]
```diff
@@ -1,148 +1,134 @@
...
+    loglik = 0.0
     for t in 2:T+1
-        n[t-1] = latent_noise(prob, t-1)
-        u[t] = prob.f(u[t - 1], params, t-1) + prob.g(u[t - 1], params, t-1) .* n[t-1] # Should this be n[t] or n[t-1]?
-        z[t] = prob.h(u[t], params, t)
+        n[t] = noise(prob.noise, t)
+        u[t] = prob.f(u[t-1], prob.params, t-1) .+ prob.g(u[t-1], prob.params, t-1) * n[t]
+        z[t] = prob.h(u[t], prob.params, t) .+ noise(prob.obs_noise, 1)
+        err = z[t] - prob.observables[t]
+        loglik += loglikelihood(err, prob.obs_noise, t)
```
Comment (ChrisRackauckas) [inline src/nonlinear.jl:130]: hmm, are there different solvers for this kind of thing?

File: cannon.jl (hunk 2 above)
Comment (wupeifan) [reply cannon.jl:31]: These are indeed tests. Guess it's Cameron's sandbox...
Speaking of tests we have a bunch of existing ones here https://github.com/HighDimensionalEconLab/DifferentiableStateSpaceModels.jl/blob/main/test/first_order_sequence.jl that maybe we can utilize. I can provide some test files as well.

File: src/noise.jl (hunk 4 above)
Comment (wupeifan) [reply src/noise.jl:34]: My two cents: if DiffEqNoiseProcess supports (i.i.d.) shock processes with a variety of distributions then we should use it.

File: cannon.jl (hunk 2 above)
Comment (ChrisRackauckas) [reply cannon.jl:31]: yeah, I would just add extra sandbox files to the tests unless they take a ton of CI time. Might as well make it more comprehensive.

File: src/noise.jl (hunk 4 above)
Comment (ChrisRackauckas) [reply src/noise.jl:34]: It has a few, and could use more. It defines it in the continuous context but those should be useable here? What kinds of shock distributions are necessary?

Comment (wupeifan) [reply src/noise.jl:34]: For those noises, I think maybe Gaussian, Student-t, Gamma, Laplace... but the key is really just i.i.d. shocks over discrete time periods.

File: integ/models.ipynb [hunk 6]
```diff
@@ -0,0 +1,512 @@
+{
```
Comment (ChrisRackauckas) [inline integ/models.ipynb:1]: It's best to keep notebooks out, and force push it out of the history 😅

File: integ/jan.jl [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    ARMA_noise, 
+    ARMA_observation, 
+    u0,
+    tspan,
+    p,
+    noise = [0.0, 0.0],
+    observables= Y,
+)
+
+solve(prob)
```
Comment (ChrisRackauckas) [inline integ/jan.jl:91]: why not make this into a test too?

File: src/noise.jl (hunk 4 above)
Comment (ChrisRackauckas) [reply src/noise.jl:34]: Ahh some of those don't have a continuous representation, so let's keep this separate.

File: src/nonlinear.jl (hunk 5 above)
Comment (cpfiffer) [reply src/nonlinear.jl:132]: Yeah, I believe there's the Kalman filter for linear forms, and various extensions that handle nonlinear models. I think.

**Comment by cpfiffer** [comment]
This should be ready for another pass -- I think I'm reasonably happy with the current status.

**Comment by wupeifan** [comment]
I was wondering what exactly should the timestamps be? For example, for a very simple AR(1) process (with time invariant `A`) if we initiate `x_0` and the shock process `epsilon`, is it
```
x_t = A * x_{t-1} + epsilon_{t}
```
for `t = 1, ..., T`?

**Comment by ChrisRackauckas** [comment]
We should make sure to match https://github.com/SciML/OrdinaryDiffEq.jl/pull/1534

**Comment by cpfiffer** [comment]
> We should make sure to match [SciML/OrdinaryDiffEq.jl#1534](https://github.com/SciML/OrdinaryDiffEq.jl/pull/1534)

I think I'm following this but I can't quite be sure -- is there another issue with more context on [SciML/OrdinaryDiffEq.jl#1534](https://github.com/SciML/OrdinaryDiffEq.jl/pull/1534)?

**Comment by cpfiffer** [comment]
I'm going to merge this for now and start spinning off issues about what we can tackle later.

**Pull request merged**